### ISSUE TITLE
Random seeds fail, every single run of remoll uses the same exact set of seeds

### ISSUE THREAD

**Issue by cameronc137** [JeffersonLab/remoll#52]
The seeds set by line 64 of remoll.cc do not affect the seeds that are used by remoll while running, in fact, the seeds never change across different runs.

What am I doing wrong? Has this been fixed already? Can the solution be added to the README?

https://github.com/JeffersonLab/remoll/blob/32234851c149ca1b8709aa2f47ec1b0012301afb/remoll.cc#L64

**Comment by wdconinc** [comment]
Master or develop?

**Comment by cameronc137** [comment]
In the develop branch, as of my last pull which was about 6 weeks ago.

**Comment by wdconinc** [comment]
I don't think anything was changes in the last six weeks but can you pull the most recent version and verify?

**Comment by cameronc137** [comment]
Now I'm getting compilation errors with the latest develop branch:

`[ 46%] Building CXX object CMakeFiles/remoll.dir/src/remollGenericDetector.cc.o
/home/cameronc/gitdir/Develop_remoll/src/remollGenericDetector.cc: In constructor ‘remollGenericDetector::remollGenericDetector(G4String, G4int)’:
/home/cameronc/gitdir/Develop_remoll/src/remollGenericDetector.cc:33: error: call of overloaded ‘to_string(G4int&)’ is ambiguous
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/basic_string.h:2604: note: candidates are: std::string std::to_string(long long int)
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/basic_string.h:2610: note:                 std::string std::to_string(long long unsigned int)
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/bits/basic_string.h:2616: note:                 std::string std::to_string(long double)
At global scope:
cc1plus: warning: unrecognized command line option "-Wno-pedantic"
make[2]: *** [CMakeFiles/remoll.dir/src/remollGenericDetector.cc.o] Error 1
make[1]: *** [CMakeFiles/remoll.dir/all] Error 2
make: *** [all] Error 2`

**Comment by wdconinc** [comment]
I'd suggest rm build dir and recompile. Is that CentOS 7?

As for seed, in the short term, try specifying the seed as remoll -s int. Or just remoll -h for other options. 

I'll look into the macro issue later today or by Monday.

**Comment by cameronc137** [comment]
It is CentOS 6.9, and recompiling gets the same error. In the meantime yes, I can probably just set the seed at runtime with -s, and I'll see if I can find the source of this compilation error.

**Comment by cameronc137** [comment]
I traced the compilation error back to commit e27c3a3318c7c580729e4c2d4a49b9db0c16c209 and the use of the std::to_string(fDetNo) command. I imagine this is an issue with my local compiler on Centos 6.9 being too old or not having an updated standard library.

**Comment by wdconinc** [comment]
Yeah, I'm hoping to keep everything compatible with ISO C++ before C++11 for at least the foreseeable future (read: until geant4.10 is required which effectively will exclude anyone on old C++) so I'm going to try to fix the to_string issue and add CentOS 6.9 to the testing matrix. I'll open a separate issue for that. I'll work on the seed in this one.

**Comment by cameronc137** [comment]
Ok, the ss.str() fix works and the latest develop branch compiles with that change. Running two comparisons at different times with different process id's still generates the same set of seeds for macro runs, so whatever is causing this problem is not new in the last 3 months.

**Comment by cameronc137** [comment]
Update: I just checked a fresh version of the develop branch, and the seeds continue to be exactly the same for every single run when running in batch mode, whether the macro command to set the seed to some other number is given or not. 

Additionally, when running just the executable as "./remoll -r INTEGER -m MACRO" even with the macro's "/remoll/seed INTEGER" used or not, it still fails to take the seed seriously.

Additionally, when going into remoll.cc and hardcoding the seed to = INTEGER at line 55, compiling, and then rerunning, even then I get the same seeds every time, even when not using the -r flag and not redeclaring the seed in the macro.

Looking in remollglobs.hh I do not see any parameters that correspond to seeds, though I have had problems with the beam energy getting over-written by the energy global parameter set in there (that's another issue I should open).

I don't know why this is happening, and all of the supported means of changing seeds seem to fail. As far as I can tell every single run of remoll uses exactly the same set of seeds every time. This is a problem for low statistics studies and is likely causing some poor results to come out. Running simulations for way longer (following the trail of the same original seed farther down the line) is ok, but it can take prohibitive amounts of time and disk space to get such results for many simulation variations.

**Comment by cameronc137** [comment]
This problem also affects the master branch.

Looking at an older branch from 2 years ago the main difference I see is that previously we had

"CLHEP::HepRandom::createInstance();                     
CLHEP::HepRandom::setTheSeed(seed);                                
remollRun::GetRun()->GetData()->SetSeed(seed); "

and now we have 

"G4Random::setTheEngine(new CLHEP::RanecuEngine()); 
G4Random::setTheSeed(seed);                        
remollRun::GetRunData()->SetSeed(seed); "

in addition to new multi-threading capabilities, which I think may also be a change that contributes to this problem - the part of remoll that reports what seeds are in use in the first place comes with some "G4WT0 or 1 > --> Event INTEGER starts with initial seeds (INTEGER,other INTEGER)." so that looks like the threading abilities have disabled our seed declarations.

**Comment by wdconinc** [comment]
Could someone please test https://github.com/JeffersonLab/remoll/tree/hotfix-random-seed (off of master) and report back? This should fix the issue. Test macro in `macros/tests/test_seed.mac`.
```
git checkout hotfix-random-seed
make -C build && build/remoll macros/tests/test_seed.mac
```
Above should generate `test_seed_0.root` with random seed (different every time you run it), `test_seed_1.root` == `test_seed_3.root` (content-wise, few bytes different file size), and `test_seed_2.root` different from 1 and 3 but the same every time you run it.

**Comment by wdconinc** [comment]
Fixed and new tag v1.1.7 released.
